GameOver Implemented

Cyc10n3z opened the pull request:
Modified the Player class to update the Player health when the kill() function is called and found a 'possible' bug that prevents the Player from losing lives at all?

Modified the update() function in PlayState to fade the camera to black and call a helper function that changes the current game state to GameOverState.

Added the GameOverState class.

Here is a screenshot of the GameOver screen when the Player reaches 0 lives:
![GameOver](https://user-images.githubusercontent.com/54049687/160227995-8db0887f-f203-4179-a923-5291d548a72d.png)

Cyc10n3z committed:
GameOver Implemented

Added the GameOverState class.

Modified the Player class to update the Player health when the kill() function is called and found a 'possible' bug that prevents the Player from losing lives at all?

Modified the update() function in PlayState to fade the camera to black and call a helper function that changes the current game state to GameOverState.

SamBumgardner commented on source/player/Player.hx:
```diff
@@ -53,8 +53,9 @@ class Player extends FlxSprite {
 
 	override function kill() {
 		reset(FlxG.width / 2, FlxG.height / 2);
-		health = maxHealth;
+		// health = maxHealth; //This line resets the players health everytime kill() is called?
 		currentPower.inUse = false;
 		currentPower.usable = true;
+		health--; // Update the players lives or health
 	}
```
With how the game functions now, custom behavior on the player's `kill()` method is basically unnecessary, right? 

If you wanted, I guess `kill()` could be the trigger that kicks off the game over sequence in the Playstate (you could take the code that's inside the `if (player.health <= 0) {` check and make it so `player.kill()` makes it happen instead). If you're not doing that, though, I think you can fully get rid of this method override and let default `kill()` just do its thing.

SamBumgardner commented on source/GameOverState.hx:
```diff
@@ -0,0 +1,25 @@
+package;
+
+import flixel.FlxG;
+import flixel.FlxSubState;
+import flixel.FlxState;
+import flixel.text.FlxText;
+
+class GameOverState extends FlxState {
+	var endMessage:FlxText;
+
+	public function new() {
+		super();
+	}
+
+	override public function create() {
+		#if FLX_MOUSE
+		FlxG.mouse.visible = true;
+		#end
+
+		super.create();
+		endMessage = new FlxText(0, 0, 0, "Game Over!", 22);
+		endMessage.screenCenter();
+		add(endMessage);
+	}
+}
```
Let's add some kind of option that lets us go back to the PlayState so players can start again if they'd like without having to reboot the game entirely

SamBumgardner reviewed (changes requested):
Base game-over functionality looks fine to me, had a couple comments to address.

I also know this is very similar to the sort of game over functionality described in the HaxeFlixel RPG tutorial. Was working through this useful to improving your understanding of working with  the engine? 

This little project is meant to help you get used to independently working through some feature of a game, and I want to make sure that you're getting that kind of value out of it. If you wanted to do more (or you think you'd get more out of doing some different feature) I'm happy to support you working through it. 

I don't want this to end up just end up feeling like a do-nothing assignment where you're ending with the same sort of understanding that you started with.

Cyc10n3z committed:
GameOver Implemented

Implemented two buttons on the Game Over screen: one button to "Try Again?" to replay the game or "Exit?" to exit the game completely.

Also modified Player.hx by removing the override function kill() and put the player position reset in the isOnScreen() check in update().

I also slightly changed or cleaned up the code I added in PlayState.hx.
